file.go

```diff
@@ -1,20 +1,135 @@
 package s3fs
 
 import (
+	"errors"
+	"fmt"
 	"io"
 	"io/fs"
+	"io/ioutil"
 	"path"
 	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
 )
 
 var (
 	_ fs.File     = (*file)(nil)
 	_ fs.FileInfo = (*fileInfo)(nil)
+	_ io.Seeker   = (*file)(nil)
 )
 
 type file struct {
+	cl     s3iface.S3API
+	bucket string
+	name   string
+
 	io.ReadCloser
-	stat func() (fs.FileInfo, error)
+	stat   func() (fs.FileInfo, error)
+	offset int64
+}
+
+func openFile(cl s3iface.S3API, bucket string, name string) (fs.File, error) {
+	out, err := cl.GetObject(&s3.GetObjectInput{
+		Key:    &name,
+		Bucket: &bucket,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	statFunc := func() (fs.FileInfo, error) {
+		return stat(cl, bucket, name)
+	}
+
+	if out.ContentLength != nil && out.LastModified != nil {
+		// if we got all the information from GetObjectOutput
+		// then we can cache fileinfo instead of making
+		// another call in case Stat is called.
+		statFunc = func() (fs.FileInfo, error) {
+			return &fileInfo{
+				name:    path.Base(name),
+				size:    *out.ContentLength,
+				modTime: *out.LastModified,
+			}, nil
+		}
+	}
+
+	return &file{
+		cl:         cl,
+		bucket:     bucket,
+		name:       name,
+		ReadCloser: out.Body,
+		stat:       statFunc,
+		offset:     0,
+	}, nil
+}
+
+func (f *file) Read(p []byte) (int, error) {
+	n, err := f.Read(p)
+	f.offset += int64(n)
+	return n, err
+}
+
+func (f *file) Seek(offset int64, whence int) (int64, error) {
+	newOffset := f.offset
+
+	stat, err := f.Stat()
+	if err != nil {
+		return 0, err
+	}
+	size := stat.Size()
+
+	switch whence {
+	case io.SeekStart:
+		newOffset = offset
+	case io.SeekCurrent:
+		newOffset += offset
+	case io.SeekEnd:
+		newOffset = size + offset
+	default:
+		return f.offset, errors.New("s3fs.file.Seek: invalid whence")
+	}
+
+	// If the position has not moved, there is no need to make a new query
+	if f.offset == newOffset {
+		return newOffset, nil
+	}
+
+	if newOffset < 0 {
+		return f.offset, errors.New("s3fs.file.Seek: seeked to a negative position")
+	}
+	if newOffset > size {
+		return f.offset, errors.New("s3fs.file.Seek: seeked to a position exceeding size")
```

Comment:
`bytes.Reader` accepts this case, it only returns the desired offset - https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/bytes/reader.go;l=116
it would return EOF on the next read. I believe we should do the same. We can remove this condition and change line 107
```go
if newOffset == size {
```
to 
```go
if newOffset >= size {
```

we are missing a test case for this, it would be worth adding one

Reply:
Changed, added tests for both == and >=.